### DISCUSSION TITLE
Add a new Safe Mode: All the memory editors of PCSX-Redux are disabled while the core is running

### DISCUSSION THREAD

**Discussion by ghost** [grumpycoders/pcsx-redux#1089] [Ideas]
I recall that the debugger of PCSX2 1.6.0 disallows the user to modify the memory of the emulated PS2 rom while the core is running.

This means that the viewed memory in the debugger window of PCSX2 can **NOT** be edited and the "Assemble Opcode" context menu command is disabled while the core is running.

In order to edit the memory or/and use the "Assemble Opcode" context menu command the core must be paused by clicking on the "Break" button on the top left corner of the debugger window of PCSX2.

This is a safe restriction because modifying the memory while the core is running can be dangerous and yield to unexpected behaviours up to crashing the debugger or the emulator!

Do you hate to see new open issues reports? Then I highly recommend you to give this feature request a high priority and implement it soon so PCSX-Redux users like me won't open new issues and complain because they modified the memory of the game while the core was running and PCSX-Redux crashed as a result!

PCSX-Redux, unlike PCSX2, allows the user to modify the memory of the emulated PS1 rom with any of it's memory editors even while the core is running and I don't think that this is a good idea at all.

What do you really think might happen if any memory of any game is changing by any memory editor while the core is running? Think about it.

I both suggest, propose and even highly recommend to at least add a new safe mode, which is enabled by default, that disables all memory editors while the core is running.

Of course that if the core was paused by either pressing F6 to pause emulation or the Pause button in the Assembly window or tab was clicked then all memory editors can still be used to modify any memory that the user wishes and desires to but until the core is resumed and all the memory editors are disabled again while safe mode is enabled.

If the user attempts to use any memory editor to modify any memory while both safe mode is enabled and the core is running then an error message box appears:

> You can not modify the memory while the core is running. Either pause emulation (RECOMMENDED) or disable safe mode (NOT RECOMMENDED)

If the user wants to modify the memory with any memory editor even while the core is running then the user must disable safe mode in order to do so but show him or her a proper warning message that the action to disable safe mode is not recommended and pausing the core or emulation instead is more recommended. Use it in your own risk!

When the "Assemble" command will be added to the context menu of the Assembly window or tab as feature requested in [https://github.com/grumpycoders/pcsx-redux/discussions/1063](url)

Then make sure that the "Assemble" context menu command shows a proper error message box while both safe mode is enabled and the core is running because changing MIPS instructions is also modifying memory for everything!

Otherwise if either safe mode is disabled or the core is **NOT** running, that is paused, then the Assemble context menu command will do what it supposes to do as described in the "Assemble" feature request which is linked above.

Right now if the user doesn't want to edit the memory of the PS1 rom that he or she is playing while the core is running with the belief that this is an unsafe act then these users must remember to pause the emulation by either pressing F6 or click on the Pause button in the Assembly window or tab, edit the memory with any memory editor of PCSX-Redux and then resume the game by either pressing F5 or clicking on the Resume button in the Assembly window or tab.

But if the user forgets to pause emulation before editing the PS1 rom's memory then the user might experience unexpected behaviour that might result in PCSX-Redux crash!

Think about it and take care!

A temporary solution and workaround to this problem is to just enable the "Use raw input for mouse" in the Controls window or tab and keep it enabled all the time or at least until the safe mode feature request is added to PCSX-Redux in the future.

While the "Use raw input for mouse" is enabled the PCSX-Redux user can **NOT** access any window and tab, including all the memory editors of PCSX-Redux and thus he or she can **NOT** edit the memory of the emulated PS1 rom while the core is running.

While the "Use raw input for mouse" is enabled the user is free to access any window and tab but only when the emulation is paused and the core is not running.

Even though both the "Pause emulation" menu command and the Pause button of the Assembly window or tab are inaccessible while the "Use raw input for mouse" is enabled the user can still pause emulation by just pressing the F6 key on his or her keyboard so this is still alright after all.

This way the PCSX-Redux user will **never** forget and **always** remember to pause both emulation and core before editing any memory with any memory editor of PCSX-Redux because he or she will always be **forced** to pause emulation and core by pressing the F6 key on his or her keyboard before editing any memory with any memory editor of PCSX-Redux!

**Comment by wheremyfoodat** [comment]
This should be feasible to add if necessary even though it'll probably be off by default because a lot of people, including me, very often edit memory that's safe to edit while the emulator is running.

Is there something that makes this so necessary though? Can't you alternatively just stop typing on the editor?

Sadly the raw mouse workaround isn't very suitable here, because most people don't know or enjoy navigating the GUI without a cursor.

**Reply by ghost** [reply]
> This should be feasible to add if necessary even though it'll probably be off by default because a lot of people, including me, very often edit memory that's safe to edit while the emulator is running.

This is just an idea, a suggestion and a proposal. Do what you think is best to do. I am just telling you what I remember from my experience with the debugger of PCSX2 1.6.0, that's all.

Perhaps you know better than me and the developers of the PCSX2 debugger about the answer to the question whether or not this is okay to edit game memory while the core is running.

Because you think different than the developers of PCSX2 Debugger I have asked the developers of PCSX2 Debugger why the PS2 memory can not be edited and the "Assemble Opcode" context menu is disabled while the core is running [here](https://github.com/PCSX2/pcsx2/discussions/7510).

Perhaps you can prove them wrong!

Anything has both **pros** and **cons** and this is the reason that there should be a checkbox for every feature so the user can interact with these checkboxes to enable/disable each feature.

There are always times that this is good to enable a feature and there are always times that this is good to disable a feature.

The safe mode should be a checkbox in the GUI but NOT hardcoded behind the scenes of course NOT!

In general hardcoding is bad practice and should be avoided where possible!

Think how awful it would have been if "Use raw input for mouse" was hardcoded and always enabled and the user couldn't disable it at all?

This is exactly what checkboxes are for after all.

If you want it unchecked by default then so be it.

Users who will find safe mode useful for them will check it and these who will find it useless, even harmful, will leave it unchecked.

The thing is the time, resources and effort that will take to add this checkbox to the GUI and implement this feature.

Of course that you don't have a time limit or a deadline to finish working on this feature.

You can finish working on this feature whenever you want and comfortable to you.

I just want to let you know that I am appreciating your hard work on this excellent PlayStation emulator and I am grateful!

> Is there something that makes this so necessary though?

This feature is of course NOT necessary and this means that this is NOT so important when this feature will be implemented because the raw mouse workaround can be used meanwhile.

> Can't you alternatively just stop typing on the editor?

I don't understand your question. Why to stop typing on the editor? If there is a parameter in the memory of the PS1 rom that I am playing that I want to increase or decrease then I would like to type on the editor to do this. I just fear that the parameter is both constantly and repeatedly read by a `lw` MIPS instruction, for example, before I finish typing all the new 8 hexadecimal digits in the memory editor. That's why I prefer to pause emulation before typing so the game engine won't use a wrong parameter that might break the game. For example the parameter was read by a `lw` MIPS instruction after I typed four hex digits but before I typed the fifth hex digit in the memory editor and now the operand register has a wrong value that might cause the PC to reach a code path that should not be followed which throws an exception and breaks the game or something.

The debugger of PCSX2 is safe because all memory, including executable code, can not be edited while the core is running and now you know why according to the above example.

> Sadly the raw mouse workaround isn't very suitable here, because most people don't know or enjoy navigating the GUI without a cursor.

Agreed and this is the reason that the safe mode has to be added to PCSX-Redux so no one will ever have to use the raw mouse workaround at all.
